pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -312,8 +313,29 @@ public BrokerResponse handleRequest(JsonNode request, @Nullable RequesterIdentit
     long routingEndTimeNs = System.nanoTime();
     _brokerMetrics.addPhaseTiming(rawTableName, BrokerQueryPhase.QUERY_ROUTING, routingEndTimeNs - routingStartTimeNs);
 
+    // Set timeout in the requests
+    long timeSpentMs = TimeUnit.NANOSECONDS.toMillis(routingEndTimeNs - compilationStartTimeNs);
+    // Remaining time in milliseconds for the server query execution
+    // NOTE: Use the max of offline table remaining time and realtime table remaining time for hybrid use case if the
+    //       timeout for them are not configured to be the same. Server side will have different remaining time set for
+    //       each table type, and broker should wait for both types to return.
+    long remainingTimeMs = 0;
+    try {
+      if (offlineBrokerRequest != null) {
+        remainingTimeMs = setQueryTimeout(offlineTableName, offlineBrokerRequest.getQueryOptions(), timeSpentMs);
+      }
+      if (realtimeBrokerRequest != null) {
+        remainingTimeMs = Math.max(remainingTimeMs,
```

Comment:
Configuring the two tables to have different timeout values seems to be an admin error, given that pinot treats the table configs independently. Under the given constraints, we can either use `Math.max` or `Math.min`, I suppose. It is worth adding a comment that this is an arbitrary choice. Perhaps `max` is better if we dont want queries suddenly timing out or having partial results due to stricter time limits.

Reply:
IMO, it could be very rare but is legit for some use cases to configure offline part to have lower timeout than the realtime part because in general offline data has better performance. This is not arbitrary choice though. To get the full result, broker has to wait for both parts done before merging the result.

Reply:
Updated the comments so that it is more clear